src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs

```diff
@@ -48,24 +62,49 @@ protected void Start()
         protected void UpdateLastAccessTime()
             => _timeSinceLastAccess = SharedStopwatch.StartNew();
 
-        protected async Task WaitForIdleAsync(IExpeditableDelaySource expeditableDelaySource)
+        protected bool Paused
         {
-            while (true)
+            get
             {
-                if (CancellationToken.IsCancellationRequested)
+                lock (_gate)
+                    return _paused_doNotAccessDirectly;
+            }
+
+            set
+            {
+                lock (_gate)
                 {
-                    return;
+                    // We should never try to transition from paused state to paused state.  That would indicate we
+                    // missed some resume call, or that the pause-notification are not serialized.  Note: we cannot make
+                    // the opposite assertion.  We start in the resumed state, and we might then get a call to resume if
+                    // we were started while in the *middle* of some global operation.
+                    if (value)
+                        Contract.ThrowIfTrue(_paused_doNotAccessDirectly);
+
+                    _paused_doNotAccessDirectly = value;
                 }
 
+                // Let subclasses know we're paused so they can change what they're doing accordingly.
+                if (value)
+                    OnPaused();
+            }
+        }
+
+        protected async Task WaitForIdleAsync(IExpeditableDelaySource expeditableDelaySource)
+        {
+            while (true)
+            {
+                this.CancellationToken.ThrowIfCancellationRequested();
+
+                // If we're not paused, and enough time has elapsed, then we're done.  Otherwise, ensure we wait at
+                // least s_minimumDelay and check again in the future.
                 var diff = _timeSinceLastAccess.Elapsed;
-                if (diff >= BackOffTimeSpan)
-                {
+                if (!ShouldWaitForIdle())
```

Comment:
@mavasani This is the core part of the change.  Instead of just saying "hey, enough time has passed, start doing your processing again" now we say "hey, even if enough time has passed, ensure we don't start doing work if we're also paused".

And the reverse is also true.  Previously, if we were paused and became unpaused, we would start up work, even if not enough time had passed since hte "last access time" had elapsed.

Basically, we had two separate checks, in two different places, and they were not coordinated.  Now, it's centralized.  In order to run solution crawler work, we need to be both non-paused *and* have seen enough time pass.

Reply:
Thanks! Sounds good to me.